---
title: "run_admixture.Rmd"
author: "SDS"
date: "Last compiled on `r Sys.Date()`"
output: 
  html_document:
    toc: true
    toc_float: true
    df_print: kable
    code_folding: hide
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(
  root.dir =
    "/Users/slacksa/OneDrive - The University of Colorado Denver")

library(tidyverse)

```

# **Overview**

RMarkdown runs Admixture using 1000G as reference. This report is not intended
to be knit - just to be run chunk by chunk.

Requires PLINK1.9, PLINK2, bcftools, and ADMIXTURE.

TO NOTE: this report and many/all of the scripts it calls assume steps 0-1 and
0-2 have been run as shown and using 1000G. If using another reference, will
need to update/remove hardcoding in the helper scripts.


# **Run Admixture**

### Steps 0-1 & 0-2

Skipping scripts 0-1 and 0-2 - already run by Shane in ancestry_estimation.qmd.

### Step 1-1

Use data after first step of pre-imputation QC in imputation repo. If final
pre_qc file was lifted over, use file right before liftover - pre_qc_hwe.

```{zsh engine.opts='-i', eval = F}
mamba activate bcftools
code_dir="/Users/slacksa/repos/tools/admixture"

bash "$code_dir"/1-1_get_comm_variants.sh \
  -r ${RKJCOLLAB}/Immunogenetics_T1D/genetics/teddy_r01/1000genomesPCA/qc/1000G.chr1.qc.pruned \
  -s ${RKJCOLLAB}/DAISY/genetics/daisy_ask_genetics/tm_r3_imp/full_qc/pre_qc/pre_qc_hwe \
  -c "$code_dir" \
  -o ${RKJCOLLAB}/DAISY/genetics/daisy_ask_genetics/admixture

```

Note: multi-allelic warning at first merge (by-chr merge of reference) -
for now, just removing these variants but could revisit in future.

17799 overlapping variants for admixture.

### Step 1-2

```{zsh engine.opts='-i', eval = F}
mamba activate bcftools
code_dir="/Users/slacksa/repos/tools/admixture"

bash "$code_dir"/1-2_merge.sh \
  -o ${RKJCOLLAB}/DAISY/genetics/daisy_ask_genetics/admixture

```


17797 variants for admixture after flipping and excluding variants that couldn't
be flipped.

### Step 1-3

```{zsh engine.opts='-i', eval = F}
mamba activate bcftools
code_dir="/Users/slacksa/repos/tools/admixture"

bash "$code_dir"/1-3_prune.sh \
  -o ${RKJCOLLAB}/DAISY/genetics/daisy_ask_genetics/admixture

```

16561 variants after pruning.

### Step 2-1

This step is slow and may be skipped if already done for reference population.
Use the K selected by CV in step 2-2.

*Notes from SDS testing: K=1: 0.68715, K=2: 0.64285.*

```{zsh engine.opts='-i', eval = F}
# mamba activate bcftools
# code_dir="/Users/slacksa/repos/tools/admixture"
# 
# bash "$code_dir"/2-1_admixture_cv.sh \
#   -o ${RKJCOLLAB}/DAISY/genetics/daisy_ask_genetics/admixture

```


### Step 2-2


```{zsh engine.opts='-i', eval = F}
mamba activate bcftools
code_dir="/Users/slacksa/repos/tools/admixture"

bash "$code_dir"/2-2_admixture.sh \
  -o ${RKJCOLLAB}/DAISY/genetics/daisy_ask_genetics/admixture \
  -k 5

```

Note this step took about 30 minutes to run (with 1 thread for admixture).

### Step 2-3

*TODO: this step has more explicit paths than other steps, but doesn't need too.*
*Update to make consistent when revisit.*


```{zsh engine.opts='-i', eval = F}
code_dir="/Users/slacksa/repos/tools/admixture"

Rscript "$code_dir"/2-3_get_eur.R \
  --q-ref ${RKJCOLLAB}/DAISY/genetics/daisy_ask_genetics/admixture/ancestry_estimation/1000G_ref.5.Q \
  --demo-ref ${RKJCOLLAB}/Immunogenetics_T1D/raw/genetic_maps/1000genomes-phase3/integrated_call_samples_v3.20130502.ALL.panel \
  --plink-ref ${RKJCOLLAB}/DAISY/genetics/daisy_ask_genetics/admixture/ancestry_estimation/1000G_ref \
  --q-study ${RKJCOLLAB}/DAISY/genetics/daisy_ask_genetics/admixture/ancestry_estimation/study.5.Q \
  --plink-study ${RKJCOLLAB}/DAISY/genetics/daisy_ask_genetics/admixture/ancestry_estimation/study

```


# **Session Info**

R Session info:

```{r session_info}
sessionInfo()
```